### PR TITLE
refactor: APIs returns ghost users when user is deleted

### LINKED ISSUE #3356: 当用户被删除时，页面和附件的发布者没有被设置为 Ghost
### 是什么版本出现了此问题？

main

### 使用的什么数据库？

PostgreSQL

### 使用的哪种方式部署？

Docker Compose

### 在线站点地址

_No response_

### 发生了什么？

如图：

image

此时如果访问这个页面，会出现异常（文章也一样，因为都取了发布者的字段）：

image

附件的上传者也没有更新：

image

/kind bug
/area core
/milestone 2.3.x

### 相关日志输出

_No response_

### 附加信息

_No response_

### PR DESCRIPTION
**Pull request by guqing**
#### What type of PR is this?
/kind improvement
/area core
/milestone 2.3.x

#### What this PR does / why we need it:
用户被删除时关联到的用户返回 ghost 用户信息

#### Which issue(s) this PR fixes:

Fixes #3356

#### Special notes for your reviewer:
/cc @halo-dev/sig-halo 
#### Does this PR introduce a user-facing change?
```release-note
用户被删除时关联到的用户返回 ghost 用户信息
```

### CODE REVIEW

File: src/main/java/run/halo/app/core/extension/service/UserService.java [hunk 1]
```diff
@@ -10,6 +10,8 @@ public interface UserService {
 
     Mono<User> getUser(String username);
 
+    Mono<User> userOrGhost(String username);
```
Comment (JohnNiang) [inline src/main/java/run/halo/app/core/extension/service/UserService.java:13]: 方法名建议更改为：`getUserOrGhost`。

**Review by JohnNiang — CHANGES_REQUESTED** [review]

File: src/main/java/run/halo/app/content/comment/CommentServiceImpl.java [hunk 2]
```diff
@@ -153,6 +153,7 @@ private Mono<ListedComment> toListedComment(Comment comment) {
 
     private Mono<OwnerInfo> getCommentOwnerInfo(Comment.CommentOwner owner) {
         if (User.KIND.equals(owner.getKind())) {
+
```
Comment (JohnNiang) [inline src/main/java/run/halo/app/content/comment/CommentServiceImpl.java:156]:
```suggestion
```

File: src/main/java/run/halo/app/core/extension/service/UserServiceImpl.java [hunk 3]
```diff
@@ -34,6 +35,12 @@ public Mono<User> getUser(String username) {
         return client.get(User.class, username);
     }
 
+    @Override
+    public Mono<User> getUserOrGhost(String username) {
+        return client.fetch(User.class, username)
+            .switchIfEmpty(Mono.defer(() -> client.fetch(User.class, GHOST_USER_NAME)));
```
Comment (JohnNiang) [inline src/main/java/run/halo/app/core/extension/service/UserServiceImpl.java:41]: 这里建议调用 `client#get`。这里得假设 ghost 用户永远存在。

**Review by JohnNiang — CHANGES_REQUESTED** [review]

File: src/main/java/run/halo/app/core/extension/service/UserServiceImpl.java (hunk 3 above)
Comment (guqing) [reply src/main/java/run/halo/app/core/extension/service/UserServiceImpl.java]: good catch

File: src/main/java/run/halo/app/theme/finders/impl/CommentFinderImpl.java [hunk 4]
```diff
@@ -107,7 +106,7 @@ private Mono<OwnerInfo> getOwnerInfo(Comment.CommentOwner owner) {
         if (Comment.CommentOwner.KIND_EMAIL.equals(owner.getKind())) {
             return Mono.just(OwnerInfo.from(owner));
         }
-        return client.fetch(User.class, owner.getName())
+        return userService.getUserOrGhost(owner.getName())
             .map(OwnerInfo::from)
             .defaultIfEmpty(OwnerInfo.ghostUser());
```
Comment (JohnNiang) [inline src/main/java/run/halo/app/theme/finders/impl/CommentFinderImpl.java:111]: 建议移除 defaultIfEmpty。

File: src/main/java/run/halo/app/content/comment/ReplyServiceImpl.java [hunk 5]
```diff
@@ -132,7 +130,7 @@ private Mono<ListedReply> toListedReply(Reply reply) {
     private Mono<OwnerInfo> getOwnerInfo(Reply reply) {
         Comment.CommentOwner owner = reply.getSpec().getOwner();
         if (User.KIND.equals(owner.getKind())) {
-            return client.fetch(User.class, owner.getName())
+            return userService.getUserOrGhost(owner.getName())
                 .map(OwnerInfo::from)
                 .switchIfEmpty(Mono.just(OwnerInfo.ghostUser()));
```
Comment (JohnNiang) [inline src/main/java/run/halo/app/content/comment/ReplyServiceImpl.java:135]: 建议移除 switchIfEmpty。

**Review by JohnNiang — CHANGES_REQUESTED** [review]

**Review by JohnNiang — APPROVED** [review]
/approve

**Review by ruibaby — APPROVED** [review]
/lgtm

**Pull request merged**